tests/integration/test_bundle.py

```diff
@@ -0,0 +1,148 @@
+
+    await _test_download_external_bundle_success()
+    await _test_download_external_bundle_http_error()
+
+
+async def _test_bundle_accessors(adcm_client: ADCMClient) -> None:
+    bundle = await adcm_client.bundles.get(name__eq="cluster_with_license")
+    assert isinstance(bundle, Bundle)
+    assert bundle.name == "cluster_with_license"
+
+    with pytest.raises(ObjectDoesNotExistError):
+        await adcm_client.bundles.get(name__eq="fake_bundle")
+
+    assert not await adcm_client.bundles.get_or_none(name__eq="fake_bundle")
+    assert isinstance(await adcm_client.bundles.get_or_none(name__contains="cluster_with"), Bundle)
+
+    bundles_list = await adcm_client.bundles.list()
+    bundles_all = await adcm_client.bundles.all()
+    assert isinstance(bundles_list, list)
+    assert len(bundles_all) == len(bundles_list) == len(os.listdir(BUNDLES))
+
+    bundles_list = await adcm_client.bundles.list(query={"limit": 2, "offset": 1})
+    assert isinstance(bundles_list, list)
+    assert len(bundles_list) == 2
+
+    bundles_list = await adcm_client.bundles.list(query={"offset": len(os.listdir(BUNDLES)) + 1})
+    assert isinstance(bundles_list, list)
+    assert len(bundles_list) == 0
+
+    async for b in adcm_client.bundles.iter(name__icontains="cluster"):
+        assert isinstance(b, Bundle)
+        assert "cluster" in b.name.lower()
+
+    assert len(await adcm_client.bundles.filter(name__icontains="cluster")) < len(os.listdir(BUNDLES))
+
+
+async def _test_bundle_properties(adcm_client: ADCMClient) -> None:
+    bundle = await adcm_client.bundles.get(name__eq="cluster_with_license")
+    assert bundle.name == "cluster_with_license"
+    assert bundle.license.state == "accepted"
+    assert "LICENSE AGREEMENT" in bundle.license.text
+    assert bundle.version == "2.0"
+    assert bundle.signature_status == "absent"
+    assert bundle.edition == "enterprise"
+
+    await bundle.license.accept()
+    await bundle.refresh()
+    assert bundle.license.state == "accepted"
+
+
+async def _test_download_external_bundle_success() -> None:
+    url = UrlPath("https://correctaddress.com")
+
+    with patch("httpx.AsyncClient.get", new_callable=AsyncMock) as mock_get:
+        mock_response = AsyncMock()
+        mock_response.ahref = httpx.Response(status_code=200, content=b"bundle content")
+        mock_get.return_value = mock_response
+        await BundleRetriever().download_external_bundle(url)
+
+        mock_get.assert_called_once_with(str(url))
+
+
+async def _test_download_external_bundle_http_error() -> None:
+    url = UrlPath("https://wrongaddress.com")
+
+    response_mock = AsyncMock()
+    response_mock.status_code = 404
+
+    with patch("httpx.AsyncClient.get", side_effect=OperationError, return_value=response_mock) as mock_get:
+        with pytest.raises(OperationError):
+            await BundleRetriever().download_external_bundle(url)
+
+        mock_get.assert_called_once_with(url)
```

Comment:
it wasn't the idea for mocking retriever

this implementation is unit-test like and tries to check that in default retriever a request is made

but what's important for testing client's API is checking that budle.create will call retriever when string is passed as source argument

because we are testing client's API behaviour, not its internals

Reply:
changed the test